File: apps/mobile/GuiltyPartyCompanion/InvitationDecoder.swift
import Foundation

enum InvitationDecoder {
    private static let requiredKeys: Set<String> = [
        "version",
        "session_id",
        "pairing_code",
        "expires_at_unix_ms",
        "gameplay_language"
    ]
    private static let sessionPattern = try! NSRegularExpression(
        pattern: "^[A-Za-z0-9_-]{16,128}$"
    )

    static func decode(
        _ rawValue: String,
        nowUnixMilliseconds: Int64 = Int64(Date().timeIntervalSince1970 * 1_000)
    ) throws -> Invitation {
        let value = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard value.count <= 1_024,
              value.range(of: "^GP1\\.[A-Za-z0-9_-]+$", options: .regularExpression) != nil
        else {
            throw SessionModelError.invalidInvitation
        }

        let encoded = String(value.dropFirst(4))
        guard encoded.count % 4 != 1 else {
            throw SessionModelError.invalidInvitation
        }
        let standard = encoded
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let padding = String(repeating: "=", count: (4 - standard.count % 4) % 4)
        guard let data = Data(base64Encoded: standard + padding),
              data.count <= 768,
              let object = try? JSONSerialization.jsonObject(with: data),
              let dictionary = object as? [String: Any],
              Set(dictionary.keys) == requiredKeys,
              let transfer = try? JSONDecoder().decode(GPV1InvitationTransfer.self, from: data),
              sessionPattern.firstMatch(
                in: transfer.sessionId.value,
                range: NSRange(transfer.sessionId.value.startIndex..., in: transfer.sessionId.value)
              ) != nil
        else {
            throw SessionModelError.invalidInvitation
        }

        guard transfer.expiresAtUnixMs > nowUnixMilliseconds else {
            throw SessionModelError.expiredInvitation
        }
        return Invitation(
            sessionID: transfer.sessionId.value,
            pairingProof: transfer.pairingCode,
            expiresAtUnixMilliseconds: transfer.expiresAtUnixMs,
            gameplayLanguage: transfer.gameplayLanguage.value
        )
    }
}

File: apps/mobile/android/app/src/main/java/com/guiltyparty/companion/InvitationDecoder.kt
package com.guiltyparty.companion

import guiltyparty.contracts.v1.GPJsonValue
import guiltyparty.contracts.v1.GPV1InvitationTransfer
import java.util.Base64

object InvitationDecoder {
    private val requiredKeys = setOf(
        "version",
        "session_id",
        "pairing_code",
        "expires_at_unix_ms",
        "gameplay_language",
    )
    private val invitationPattern = Regex("^GP1\\.[A-Za-z0-9_-]+$")
    private val sessionPattern = Regex("^[A-Za-z0-9_-]{16,128}$")

    fun decode(
        rawValue: String,
        nowUnixMs: Long = System.currentTimeMillis(),
    ): Invitation {
        val value = rawValue.trim()
        if (value.length > 1_024 || !invitationPattern.matches(value)) {
            throw CompanionFailure(FailureKind.INVALID_INVITATION)
        }
        val encoded = value.drop(4)
        if (encoded.length % 4 == 1) {
            throw CompanionFailure(FailureKind.INVALID_INVITATION)
        }
        val bytes = runCatching { Base64.getUrlDecoder().decode(encoded) }.getOrNull()
            ?: throw CompanionFailure(FailureKind.INVALID_INVITATION)
        if (bytes.size > 768) {
            throw CompanionFailure(FailureKind.INVALID_INVITATION)
        }
        val json = runCatching { ContractJson.parse(ContractJson.decodeUtf8(bytes)) }.getOrNull()
            ?: throw CompanionFailure(FailureKind.INVALID_INVITATION)
        val members = (json as? GPJsonValue.ObjectValue)?.members
            ?: throw CompanionFailure(FailureKind.INVALID_INVITATION)
        if (members.keys != requiredKeys) {
            throw CompanionFailure(FailureKind.INVALID_INVITATION)
        }
        val transfer = runCatching { GPV1InvitationTransfer.from(json) }.getOrNull()
            ?: throw CompanionFailure(FailureKind.INVALID_INVITATION)
        if (!sessionPattern.matches(transfer.sessionId.value)) {
            throw CompanionFailure(FailureKind.INVALID_INVITATION)
        }
        if (transfer.expiresAtUnixMs <= nowUnixMs) {
            throw CompanionFailure(FailureKind.EXPIRED_INVITATION)
        }
        return Invitation(
            sessionId = transfer.sessionId.value,
            pairingProof = transfer.pairingCode,
            expiresAtUnixMs = transfer.expiresAtUnixMs,
            gameplayLanguage = transfer.gameplayLanguage.value,
        )
    }
}
